README.md
## Description

Builds an edge map using canny edge detection.

This is [a port of the VapourSynth plugin TEdgeMask](https://github.com/dubhater/vapoursynth-tedgemask).

### Requirements:

- AviSynth 2.60 / AviSynth+ 3.4 or later

- Microsoft VisualC++ Redistributable Package 2022 (can be downloaded from [here](https://github.com/abbodi1406/vcredist/releases))

### Usage:

```
vsTEdgeMask (clip, float "threshY", float "threshU", float "threshV", int "type", int "link", float "scale", int "y", int "u", int "v", int "opt")
```

### Parameters:

- clip\
    A clip to process. It must have planar format, 8..16 bit integer sample type, and subsampling ratios of at most 2.

- threshY, threshU, threshV\
    Sets the magnitude thresholds.\
    If over this value then a sample will be considered an edge, and the output pixel will be set to the maximum value allowed by the format. Otherwise the output pixel will be set to 0.\
    Set this to 0 to output a magnitude mask instead of a binary mask.\
    Default: threshY = threshU = threshV = 8.0.

- type\
    Sets the type of first order partial derivative approximation that is used.\
    1: 2 pixel.\
    2: 4 pixel.\
    3: Same as type = 1.\
    4: Same as type = 2.\
    5: 6 pixel (Sobel operator).\
    Default: 2.

- link\
    Specifies whether luma to chroma linking, no linking, or linking of every plane to every other plane is used.\
    0: No linking. The three edge masks are completely independent.\
    1: Luma to chroma linking. If a luma pixel is considered an edge, the corresponding chroma pixel is also marked as an edge.\
    2: Every plane to every other plane. If a pixel is considered an edge in any plane, the corresponding pixels in all the other planes are also marked as edges.\
    This parameter has no effect when clip has only one plane, when any plane's threshold is 0, or when some planes are not processed.\
    This parameter can only be 0 or 2 when clip is RGB.\
    Default: 2 when clip is RGB, otherwise 1.

- scale\
    If the output is a magnitude mask (threshold is 0), it is scaled by this value.\
    Note that in TEMmod this parameter had three different, undocumented default values for the different mask types, which made it difficult to use the parameter without reading the source code.\
    Default: 1.0.

- y, u, v\
    Planes to process.\
    1: Return garbage.\
    2: Copy plane.\
    3: Process plane. Always process planes when clip is RGB.\
    Default: y = u = v = 3.

- opt\
    Sets which cpu optimizations to use.\
    -1: Auto-detect.\
    0: Use C++ code.\
    1: Use SSE2 code.\
    2: Use AVX2 code.\
    3: Use AVX512 code.\
    Default: -1.

### Building:

- Windows\
    Use solution files.

- Linux
    ```
    Requirements:
        - Git
        - C++11 compiler
        - CMake >= 3.16
    ```
    ```
    git clone https://github.com/Asd-g/AviSynth-vsTEdgeMask && \
    cd AviSynth-vsTEdgeMask && \
    mkdir build && \
    cd build && \
    cmake .. && \
    make -j$(nproc) && \
    sudo make install
    ```


CHANGELOG.md
##### 1.0.1:
    Added AVX2 and AVX512 code.

##### 1.0.0:
    Initial release.


src/vsTEdgeMask.h
#pragma once

#include "avisynth.h"

enum MaskTypes
{
    TwoPixel = 1,
    FourPixel = 2,
    SixPixel = 5
};

template<typename PixelType, MaskTypes type, bool binarize>
void detect_edges_sse2(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template<typename PixelType, MaskTypes type, bool binarize>
void detect_edges_avx2(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template<typename PixelType, MaskTypes type, bool binarize>
void detect_edges_avx512(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;

class vsTEdgeMask : public GenericVideoFilter
{
    int64_t _threshold[3];
    int _link;
    double _scale;
    int process[3];
    bool has_at_least_v8;

    void (*detect_edges[3])(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
    void (*link_planes)(uint8_t* __restrict dstp18, uint8_t* __restrict dstp28, uint8_t* __restrict dstp38, int stride1, int stride2, int width, int height, int pixel_max) noexcept;

public:
    vsTEdgeMask(PClip _child, double threshY, double threshU, double threshV, int type, int link, float scale, int y, int u, int v, int opt, IScriptEnvironment* env);
    PVideoFrame __stdcall GetFrame(int n, IScriptEnvironment* env);
    int __stdcall SetCacheHints(int cachehints, int frame_range)
    {
        return cachehints == CACHE_GET_MTMODE ? MT_NICE_FILTER : 0;
    }
};


src/vsTEdgeMask_AVX2.cpp
#include <memory>

#include "VCL2/vectorclass.h"
#include "vsTEdgeMask.h"


template<typename PixelType, MaskTypes type, bool binarize>
static AVS_FORCEINLINE void detect_edges_uint8_mmword_avx2(const PixelType* srcp, PixelType* __restrict dstp, int x, int stride, const Vec8i& threshold, float scale) noexcept
{
    Vec16s gx, gy;
    Vec8f divisor;

    const auto top = Vec16s().load_16uc(srcp + (x - stride));
    const auto left = Vec16s().load_16uc(srcp + (x - 1));
    const auto right = Vec16s().load_16uc(srcp + (x + 1));
    const auto bottom = Vec16s().load_16uc(srcp + (x + stride));

    if constexpr (type == TwoPixel)
    {
        gx = right - left;
        gy = top - bottom;
        divisor = 0.25f;
    }
    else if constexpr (type == FourPixel)
    {
        const auto top2 = Vec16s().load_16uc(srcp + (x - 2 * stride));
        const auto left2 = Vec16s().load_16uc(srcp + (x - 2));
        const auto right2 = Vec16s().load_16uc(srcp + (x + 2));
        const auto bottom2 = Vec16s().load_16uc(srcp + (x + 2 * stride));

        gx = 12 * (left2 - right2) + 74 * (right - left);

        gy = 12 * (bottom2 - top2) + 74 * (top - bottom);

        divisor = 0.0001f;
    }
    else if constexpr (type == SixPixel)
    {
        const auto top_left = Vec16s().load_16uc(srcp + (x - stride - 1));
        const auto top_right = Vec16s().load_16uc(srcp + (x - stride + 1));
        const auto bottom_left = Vec16s().load_16uc(srcp + (x + stride - 1));
        const auto bottom_right = Vec16s().load_16uc(srcp + (x + stride + 1));

        const auto sub_right_left = right - left;
        const auto sub_bottom_top = bottom - top;

        gx = sub_right_left + sub_right_left + (top_right - top_left) + (bottom_right - bottom_left);

        gy = sub_bottom_top + sub_bottom_top + (bottom_left - top_left) + (bottom_right - top_right);

        divisor = 1.0f;
    }

    const auto gx_gy_lo = blend16<0, 16, 1, 17, 2, 18, 3, 19, 4, 20, 5, 21, 6, 22, 7, 23>(gx, gy);
    const auto gx_gy_hi = blend16<8, 24, 9, 25, 10, 26, 11, 27, 12, 28, 13, 29, 14, 30, 15, 31>(gx, gy);

    if constexpr (binarize)
        compress_saturated(compress_saturated(madd(gx_gy_lo, gx_gy_lo) > threshold, madd(gx_gy_hi, gx_gy_hi) > threshold), zero_si256()).get_low().store(dstp + x);
    else
        compress_saturated(compress_saturated_s2u(truncatei(sqrt(to_float(madd(gx_gy_lo, gx_gy_lo)) * divisor) * scale + 0.5f),
            truncatei(sqrt(to_float(madd(gx_gy_hi, gx_gy_hi)) * divisor) * scale + 0.5f)), zero_si256()).get_low().store(dstp + x);
}


template<typename PixelType, MaskTypes type, bool binarize>
static AVS_FORCEINLINE void detect_edges_uint16_mmword_avx2(const PixelType* srcp, PixelType* __restrict dstp, int x, int stride, const Vec8f& threshold, float scale, int pixel_max) noexcept
{
    Vec8f gx, gy, divisor;

    const auto top = Vec8i().load_8us(srcp + (x - stride));
    const auto left = Vec8i().load_8us(srcp + (x - 1));
    const auto right = Vec8i().load_8us(srcp + (x + 1));
    const auto bottom = Vec8i().load_8us(srcp + (x + stride));

    if constexpr (type == TwoPixel)
    {
        gx = to_float(right - left);
        gy = to_float(top - bottom);
        divisor = 0.25f;
    }
    else if constexpr (type == FourPixel)
    {
        const auto top2 = Vec8i().load_8us(srcp + (x - 2 * stride));
        const auto left2 = Vec8i().load_8us(srcp + (x - 2));
        const auto right2 = Vec8i().load_8us(srcp + (x + 2));
        const auto bottom2 = Vec8i().load_8us(srcp + (x + 2 * stride));

        const auto sub_left2_right2 = left2 - right2;
        const auto sub_right_left = right - left;

        gx = to_float(sub_left2_right2) * 12.0f + to_float(sub_right_left) * 74.0f;

        const auto sub_bottom2_top2 = bottom2 - top2;
        const auto sub_top_bottom = top - bottom;

        gy = to_float(sub_bottom2_top2) * 12.0f + to_float(sub_top_bottom) * 74.0f;

        divisor = 0.0001f;
    }
    else if constexpr (type == SixPixel)
    {
        const auto top_left = Vec8i().load_8us(srcp + (x - stride - 1));
        const auto top_right = Vec8i().load_8us(srcp + (x - stride + 1));
        const auto bottom_left = Vec8i().load_8us(srcp + (x + stride - 1));
        const auto bottom_right = Vec8i().load_8us(srcp + (x + stride + 1));

        const auto sub_right_left = right - left;
        const auto sub_bottom_top = bottom - top;

        gx = to_float((sub_right_left + sub_right_left) + (top_right - top_left) + (bottom_right - bottom_left));

        gy = to_float(sub_bottom_top + sub_bottom_top + (bottom_left - top_left) + (bottom_right - top_right));

        divisor = 1.0f;
    }

    if constexpr (binarize)
        (compress_saturated(Vec8i((gx * gx + gy * gy) > threshold), zero_si256()) & pixel_max).get_low().store(dstp + x);
    else
    {
        auto output = truncatei(sqrt((gx * gx + gy * gy) * divisor) * scale + -32767.5f); // 0.5 for rounding and -32768 for packing

        (min(compress_saturated(output, output), pixel_max - 32768) + 32768).get_low().store(dstp + x);
    }
}


template<typename PixelType, MaskTypes type, bool binarize>
void detect_edges_avx2(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept
{
    const PixelType* srcp = reinterpret_cast<const PixelType*>(srcp8);
    PixelType* __restrict dstp = reinterpret_cast<PixelType*>(dstp8);
    stride /= sizeof(PixelType);
    dst_stride /= sizeof(PixelType);
    width /= sizeof(PixelType);

    const Vec8i threshold_epi32 = static_cast<int>(threshold64);
    const Vec8f threshold_ps = static_cast<float>(threshold64);

    // Number of pixels to skip at the edges of the image.
    const int skip = type == FourPixel ? 2 : 1;

    const int pixels_per_iteration = 16 / sizeof(PixelType);

    const int width_simd = (width - 2 * skip) / pixels_per_iteration * pixels_per_iteration + 2 * skip;

    for (int y = 0; y < skip; ++y)
    {
        memset(dstp, 0, width * sizeof(PixelType));

        srcp += stride;
        dstp += dst_stride;
    }

    for (int y = skip; y < height - skip; ++y)
    {
        memset(dstp, 0, skip * sizeof(PixelType));

        for (int x = skip; x < width_simd - skip; x += pixels_per_iteration)
        {
            if constexpr (std::is_same_v<PixelType, uint8_t>)
                detect_edges_uint8_mmword_avx2<PixelType, type, binarize>(srcp, dstp, x, stride, threshold_epi32, scale);
            else
                detect_edges_uint16_mmword_avx2<PixelType, type, binarize>(srcp, dstp, x, stride, threshold_ps, scale, pixel_max);
        }

        if (width > width_simd)
        {
            if constexpr (std::is_same_v<PixelType, uint8_t>)
                detect_edges_uint8_mmword_avx2<PixelType, type, binarize>(srcp, dstp, width - skip - pixels_per_iteration, stride, threshold_epi32, scale);
            else
                detect_edges_uint16_mmword_avx2<PixelType, type, binarize>(srcp, dstp, width - skip - pixels_per_iteration, stride, threshold_ps, scale, pixel_max);
        }

        memset(dstp + width - skip, 0, skip * sizeof(PixelType));

        srcp += stride;
        dstp += dst_stride;
    }

    for (int y = height - skip; y < height; ++y)
    {
        memset(dstp, 0, width * sizeof(PixelType));

        srcp += stride;
        dstp += dst_stride;
    }
}

template void detect_edges_avx2<uint8_t, TwoPixel, false>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template void detect_edges_avx2<uint8_t, FourPixel, false>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template void detect_edges_avx2<uint8_t, SixPixel, false>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;

template void detect_edges_avx2<uint8_t, TwoPixel, true>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template void detect_edges_avx2<uint8_t, FourPixel, true>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template void detect_edges_avx2<uint8_t, SixPixel, true>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;

template void detect_edges_avx2<uint16_t, TwoPixel, false>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template void detect_edges_avx2<uint16_t, FourPixel, false>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template void detect_edges_avx2<uint16_t, SixPixel, false>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;

template void detect_edges_avx2<uint16_t, TwoPixel, true>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template void detect_edges_avx2<uint16_t, FourPixel, true>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template void detect_edges_avx2<uint16_t, SixPixel, true>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;


src/vsTEdgeMask_AVX512.cpp
#include <memory>

#include "VCL2/vectorclass.h"
#include "vsTEdgeMask.h"


template<typename PixelType, MaskTypes type, bool binarize>
static AVS_FORCEINLINE void detect_edges_uint8_mmword_avx512(const PixelType* srcp, PixelType* __restrict dstp, int x, int stride, const Vec16i& threshold, float scale) noexcept
{
    Vec32s gx, gy;
    Vec16f divisor;

    const auto top = Vec32s().load_32uc(srcp + (x - stride));
    const auto left = Vec32s().load_32uc(srcp + (x - 1));
    const auto right = Vec32s().load_32uc(srcp + (x + 1));
    const auto bottom = Vec32s().load_32uc(srcp + (x + stride));

    if constexpr (type == TwoPixel)
    {
        gx = right - left;
        gy = top - bottom;
        divisor = 0.25f;
    }
    else if constexpr (type == FourPixel)
    {
        const auto top2 = Vec32s().load_32uc(srcp + (x - 2 * stride));
        const auto left2 = Vec32s().load_32uc(srcp + (x - 2));
        const auto right2 = Vec32s().load_32uc(srcp + (x + 2));
        const auto bottom2 = Vec32s().load_32uc(srcp + (x + 2 * stride));

        gx = 12 * (left2 - right2) + 74 * (right - left);

        gy = 12 * (bottom2 - top2) + 74 * (top - bottom);

        divisor = 0.0001f;
    }
    else if constexpr (type == SixPixel)
    {
        const auto top_left = Vec32s().load_32uc(srcp + (x - stride - 1));
        const auto top_right = Vec32s().load_32uc(srcp + (x - stride + 1));
        const auto bottom_left = Vec32s().load_32uc(srcp + (x + stride - 1));
        const auto bottom_right = Vec32s().load_32uc(srcp + (x + stride + 1));

        const auto sub_right_left = right - left;
        const auto sub_bottom_top = bottom - top;

        gx = sub_right_left + sub_right_left + (top_right - top_left) + (bottom_right - bottom_left);

        gy = sub_bottom_top + sub_bottom_top + (bottom_left - top_left) + (bottom_right - top_right);

        divisor = 1.0f;
    }

    const auto gx_gy_lo = blend32<0, 32, 1, 33, 2, 34, 3, 35, 4, 36, 5, 37, 6, 38, 7, 39, 8, 40, 9, 41, 10, 42, 11, 43, 12, 44, 13, 45, 14, 46, 15, 47>(gx, gy);
    const auto gx_gy_hi = blend32<16, 48, 17, 49, 18, 50, 19, 51, 20, 52, 21, 53, 22, 54, 23, 55, 24, 56, 25, 57, 26, 58, 27, 59, 28, 60, 29, 61, 30, 62, 31, 63>(gx, gy);

    if constexpr (binarize)
        compress_saturated(compress_saturated(select(madd(gx_gy_lo, gx_gy_lo) > threshold, Vec16i(-1), zero_si512()),
            select(madd(gx_gy_hi, gx_gy_hi) > threshold, Vec16i(-1), zero_si512())), zero_si512()).get_low().store(dstp + x);
    else
        compress_saturated(compress_saturated_s2u(truncatei(sqrt(to_float(madd(gx_gy_lo, gx_gy_lo)) * divisor) * scale + 0.5f),
            truncatei(sqrt(to_float(madd(gx_gy_hi, gx_gy_hi)) * divisor) * scale + 0.5f)), zero_si512()).get_low().store(dstp + x);
}


template<typename PixelType, MaskTypes type, bool binarize>
static AVS_FORCEINLINE void detect_edges_uint16_mmword_avx512(const PixelType* srcp, PixelType* __restrict dstp, int x, int stride, const Vec16f& threshold, float scale, int pixel_max) noexcept
{
    Vec16f gx, gy, divisor;

    const auto top = Vec16i().load_16us(srcp + (x - stride));
    const auto left = Vec16i().load_16us(srcp + (x - 1));
    const auto right = Vec16i().load_16us(srcp + (x + 1));
    const auto bottom = Vec16i().load_16us(srcp + (x + stride));

    if constexpr (type == TwoPixel)
    {
        gx = to_float(right - left);
        gy = to_float(top - bottom);
        divisor = 0.25f;
    }
    else if constexpr (type == FourPixel)
    {
        const auto top2 = Vec16i().load_16us(srcp + (x - 2 * stride));
        const auto left2 = Vec16i().load_16us(srcp + (x - 2));
        const auto right2 = Vec16i().load_16us(srcp + (x + 2));
        const auto bottom2 = Vec16i().load_16us(srcp + (x + 2 * stride));

        const auto sub_left2_right2 = left2 - right2;
        const auto sub_right_left = right - left;

        gx = to_float(sub_left2_right2) * 12.0f + to_float(sub_right_left) * 74.0f;

        const auto sub_bottom2_top2 = bottom2 - top2;
        const auto sub_top_bottom = top - bottom;

        gy = to_float(sub_bottom2_top2) * 12.0f + to_float(sub_top_bottom) * 74.0f;

        divisor = 0.0001f;
    }
    else if constexpr (type == SixPixel)
    {
        const auto top_left = Vec16i().load_16us(srcp + (x - stride - 1));
        const auto top_right = Vec16i().load_16us(srcp + (x - stride + 1));
        const auto bottom_left = Vec16i().load_16us(srcp + (x + stride - 1));
        const auto bottom_right = Vec16i().load_16us(srcp + (x + stride + 1));

        const auto sub_right_left = right - left;
        const auto sub_bottom_top = bottom - top;

        gx = to_float((sub_right_left + sub_right_left) + (top_right - top_left) + (bottom_right - bottom_left));

        gy = to_float(sub_bottom_top + sub_bottom_top + (bottom_left - top_left) + (bottom_right - top_right));

        divisor = 1.0f;
    }

    if constexpr (binarize)
        (compress_saturated(truncatei(select(Vec16fb((gx * gx + gy * gy) > threshold), Vec16f(-1.0f), zero_16f())), zero_si512()) & pixel_max).get_low().store(dstp + x);
    else
    {
        auto output = truncatei(sqrt((gx * gx + gy * gy) * divisor) * scale + -32767.5f); // 0.5 for rounding and -32768 for packing

        (min(compress_saturated(output, output), pixel_max - 32768) + 32768).get_low().store(dstp + x);
    }
}


template<typename PixelType, MaskTypes type, bool binarize>
void detect_edges_avx512(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept
{
    const PixelType* srcp = reinterpret_cast<const PixelType*>(srcp8);
    PixelType* __restrict dstp = reinterpret_cast<PixelType*>(dstp8);
    stride /= sizeof(PixelType);
    dst_stride /= sizeof(PixelType);
    width /= sizeof(PixelType);

    const Vec16i threshold_epi32 = static_cast<int>(threshold64);
    const Vec16f threshold_ps = static_cast<float>(threshold64);

    // Number of pixels to skip at the edges of the image.
    const int skip = type == FourPixel ? 2 : 1;

    const int pixels_per_iteration = 32 / sizeof(PixelType);

    const int width_simd = (width - 2 * skip) / pixels_per_iteration * pixels_per_iteration + 2 * skip;

    for (int y = 0; y < skip; ++y)
    {
        memset(dstp, 0, width * sizeof(PixelType));

        srcp += stride;
        dstp += dst_stride;
    }

    for (int y = skip; y < height - skip; ++y)
    {
        memset(dstp, 0, skip * sizeof(PixelType));

        for (int x = skip; x < width_simd - skip; x += pixels_per_iteration)
        {
            if constexpr (std::is_same_v<PixelType, uint8_t>)
                detect_edges_uint8_mmword_avx512<PixelType, type, binarize>(srcp, dstp, x, stride, threshold_epi32, scale);
            else
                detect_edges_uint16_mmword_avx512<PixelType, type, binarize>(srcp, dstp, x, stride, threshold_ps, scale, pixel_max);
        }

        if (width > width_simd)
        {
            if constexpr (std::is_same_v<PixelType, uint8_t>)
                detect_edges_uint8_mmword_avx512<PixelType, type, binarize>(srcp, dstp, width - skip - pixels_per_iteration, stride, threshold_epi32, scale);
            else
                detect_edges_uint16_mmword_avx512<PixelType, type, binarize>(srcp, dstp, width - skip - pixels_per_iteration, stride, threshold_ps, scale, pixel_max);
        }

        memset(dstp + width - skip, 0, skip * sizeof(PixelType));

        srcp += stride;
        dstp += dst_stride;
    }

    for (int y = height - skip; y < height; ++y)
    {
        memset(dstp, 0, width * sizeof(PixelType));

        srcp += stride;
        dstp += dst_stride;
    }
}

template void detect_edges_avx512<uint8_t, TwoPixel, false>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template void detect_edges_avx512<uint8_t, FourPixel, false>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template void detect_edges_avx512<uint8_t, SixPixel, false>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;

template void detect_edges_avx512<uint8_t, TwoPixel, true>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template void detect_edges_avx512<uint8_t, FourPixel, true>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template void detect_edges_avx512<uint8_t, SixPixel, true>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;

template void detect_edges_avx512<uint16_t, TwoPixel, false>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template void detect_edges_avx512<uint16_t, FourPixel, false>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template void detect_edges_avx512<uint16_t, SixPixel, false>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;

template void detect_edges_avx512<uint16_t, TwoPixel, true>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template void detect_edges_avx512<uint16_t, FourPixel, true>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
template void detect_edges_avx512<uint16_t, SixPixel, true>(const uint8_t* srcp8, uint8_t* __restrict dstp8, int stride, int dst_stride, int width, int height, int64_t threshold64, float scale, int pixel_max) noexcept;
